File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/VfsRootHvsc.java
/**
 * @file
 * @brief Implementation of VfsRoot over http://hvsc.c64.org collection
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.fs;

import android.content.Context;
import android.net.Uri;

import androidx.annotation.Nullable;

import app.zxtune.R;
import app.zxtune.fs.http.MultisourceHttpProvider;
import app.zxtune.fs.httpdir.Catalog;
import app.zxtune.fs.httpdir.HttpRootBase;
import app.zxtune.fs.hvsc.Path;

@Icon(R.drawable.ic_browser_vfs_hvsc)
final class VfsRootHvsc extends HttpRootBase implements VfsRoot {

  private static final String[] SUBDIRS = {"DEMOS/", "GAMES/", "MUSICIANS/"};

  private final Context context;

  VfsRootHvsc(VfsObject parent, Context context, MultisourceHttpProvider http) {
    super(parent, Catalog.create(context, http, "hvsc"), Path.create());
    this.context = context;
  }

  @Override
  public String getName() {
    return context.getString(R.string.vfs_hvsc_root_name);
  }

  @Override
  public String getDescription() {
    return context.getString(R.string.vfs_hvsc_root_description);
  }

  @Override
  public void enumerate(Visitor visitor) {
    for (String dir : SUBDIRS) {
      visitor.onDir(makeDir(rootPath.getChild(dir), ""));
    }
  }

  @Override
  @Nullable
  public VfsObject resolve(Uri uri) {
    return resolve(Path.parse(uri));
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/VfsRootHvsc.kt
/**
 * @file
 * @brief Implementation of VfsRoot over http://hvsc.c64.org collection
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.fs

import android.content.Context
import android.net.Uri
import app.zxtune.R
import app.zxtune.fs.http.MultisourceHttpProvider
import app.zxtune.fs.httpdir.Catalog
import app.zxtune.fs.httpdir.HttpRootBase
import app.zxtune.fs.hvsc.Path

internal class VfsRootHvsc(
    parent: VfsObject,
    private val context: Context,
    http: MultisourceHttpProvider
) : HttpRootBase(parent, Catalog.create(context, http, "hvsc"), Path.create()), VfsRoot {

    override val name: String
        get() = context.getString(R.string.vfs_hvsc_root_name)
    override val description: String
        get() = context.getString(R.string.vfs_hvsc_root_description)

    override fun enumerate(visitor: VfsDir.Visitor) = SUBDIRS.forEach { dir ->
        visitor.onDir(makeDir(rootPath.getChild(dir), ""))
    }

    override fun getExtension(id: String) = when (id) {
        VfsExtensions.ICON -> R.drawable.ic_browser_vfs_hvsc
        else -> super.getExtension(id)
    }

    override fun resolve(uri: Uri) = resolve(Path.parse(uri))

    companion object {
        private val SUBDIRS = arrayOf("DEMOS/", "GAMES/", "MUSICIANS/")
    }
}
